### PR TITLE
wip: e2e: fix reboot test flakes

### LINKED ISSUE #837: e2e: TestReboot test doesn't wait for real control plane to be restored, can cause flakes in TestSmoke
TestReboot only waits for the nodes to be ready. Based on my testing, the control plane is just the checkpointed control plane, not the actual one. When the control plane does come back, there's a point where neither is running, which can impact the TestSmoke test.

In https://github.com/kubernetes-incubator/bootkube/pull/831 I tried to wait for the API server to loose the checkpointer annotation, which showed that it can take more than 5 min (!) for the real control plane to come back up.

Output here:

https://jenkins-kube-lifecycle.prod.coreos.systems/job/bootkube-pr-e2e/716/console

Example of pods with the checkpointer annotation on them in kube-system:

```
00:10:36 [163005.176363] golang[6]: E0113 00:10:36.181749    2147 main_test.go:161] attempt 12: kube-system still has checkpoint pods=["kube-apiserver-f22bh-ip-10-8-0-153.ec2.internal" "pod-checkpointer-95fbd-ip-10-8-0-153.ec2.internal"], non-checkpoint pods=["kube-controller-manager-c5c47f48-bthqh" "kube-controller-manager-c5c47f48-fcmpz" "kube-dns-ff8f6db49-z48r4" "kube-flannel-sjfrx" "kube-flannel-trvbt" "kube-proxy-dhd99" "kube-proxy-vt9g4" "kube-scheduler-546ff4f59b-6kqqw" "kube-scheduler-546ff4f59b-7nqbl" "pod-checkpointer-95fbd"]
```

cc @colemickens

### PR DESCRIPTION
**Pull request by colemickens**
This is Eric's fix/changes from #831. Fixes #837.

### CODE REVIEW

**Comment by xiang90** [comment]
@colemickens have you tried to create a jenkins job to stress out the tests this PR indents to fix? Want to make sure this PR does fix things before merging.

File: e2e/reboot_test.go [hunk 1]
```diff
@@ -76,3 +80,35 @@ func nodesReady(c kubernetes.Interface, expectedNodes *v1.NodeList, t *testing.T
...
+			if len(pod.Annotations) > 0 && pod.Annotations[checkpointAnnotation] != "" {
+				checkpointPods = append(checkpointPods, pod.Name)
+			} else {
+				regularPods = append(regularPods, pod.Name)
+			}
+		}
+		if len(checkpointPods) > 0 {
+			sort.Strings(checkpointPods)
+			sort.Strings(regularPods)
+			return fmt.Errorf("kube-system still has checkpoint pods=%q, non-checkpoint pods=%q", checkpointPods, regularPods)
```
Comment (xiang90) [inline e2e/reboot_test.go]: this error message might super long, and not easy to read?

Comment (colemickens) [reply e2e/reboot_test.go]: ack, will clean it up

File: e2e/reboot_test.go [hunk 2]
```diff
@@ -76,3 +80,35 @@ func nodesReady(c kubernetes.Interface, expectedNodes *v1.NodeList, t *testing.T
 		return nil
 	})
 }
+
+const checkpointAnnotation = "checkpointer.alpha.coreos.com/checkpoint-of"
+
+// controlPlaneReady waits for API server availability and no checkpointed pods
+// in kube-system.
+func controlPlaneReady(c kubernetes.Interface, attempts int, backoff time.Duration) error {
+	return retry(atttempts, backoff, func() error {
```
Comment (ericchiang) [inline e2e/reboot_test.go]: s/atttempts/attempts/g

**Comment by ericchiang** [comment]
@colemickens it looks like the checkpointer always runs a checkpointed instance of itself

```
$ kubectl get pods -n kube-system -o json | \
    jq '.items[].metadata | "\(.name) \(.annotations."checkpointer.alpha.coreos.com/checkpoint-of")"' -r
kube-apiserver-bbpf2 null
kube-controller-manager-c5c47f48-7mn66 null
kube-controller-manager-c5c47f48-gm7rk null
kube-dns-ff8f6db49-xcbw7 null
kube-flannel-g2gr8 null
kube-proxy-fdtqc null
kube-scheduler-546ff4f59b-pm54v null
kube-scheduler-546ff4f59b-r55d7 null
pod-checkpointer-tl67l null
pod-checkpointer-tl67l-172.17.4.100 pod-checkpointer-tl67l
```

Maybe we should check if the `checkpointer.alpha.coreos.com/checkpoint-of` value is also a running pod?

```
parentPod := pod.Annotations[checkpointAnnotation]
foundParent := false
for _, p := range pods.Items {
    if p.Name == parentPod {
        foundParent = p.Status.Phase == "Running"
        break
    }
}
if !foundParent {
    waitingOnPods = append(waitingOnPods, pod)
}
```

**Comment by colemickens** [comment]
Code makes sense. Let me add this and make sure I understand the checkpointer behavior enough...

**Comment by ericchiang** [comment]
@colemickens can you squash before pushing too?

**Comment by colemickens** [comment]
@ericchiang Do you know if this is actually expected behavior? I'm looking at the docs for checkpointer, and based on my reading, it sounds like the checkpointed checkpointer instance should have been removed...

>The pod checkpoint will also checkpoint itself to the disk to handle the absence of the API server. After a node reboot, the on-disk pod-checkpointer will take over the responsibility. Once it reaches the API server and finds out that it's no longer being scheduled, it will clean up itself.

https://github.com/kubernetes-incubator/bootkube/tree/master/cmd/checkpoint

**Comment by colemickens** [comment]
Per the discussion in Slack, this is expected behavior. The docs for bootkube have been updated to clarify this behavior as well: https://github.com/kubernetes-incubator/bootkube/pull/856

Running this again locally. Will ping the thread when it's looking good.

**Comment by colemickens** [comment]
@xiang90 I do have tests setup locally, as well as a job in Jenkins. I should've labeled this as WIP until it was ready to merge.

File: e2e/main_test.go [hunk 3]
```diff
@@ -157,6 +158,7 @@ func retry(attempts int, delay time.Duration, f func() error) error {
 		}
 
 		if i < attempts-1 {
+			glog.Errorf("attempt %d: %v", i, err)
```
Comment (ericchiang) [inline e2e/main_test.go]: I think this turned out too spammy. Maybe remove?

Comment (colemickens) [reply e2e/main_test.go]: Yes. Will remove.

**Comment by colemickens** [comment]
Yes, the TestReboot passed. Looks like they flaked on TestSmoke (next candidate probably?).

Will go ahead and squash, merge, push. I'm going to assume you're okay with me merging, even if there is red from TestSmoke (or something else, other than TestReboot) flaking.

**Comment by colemickens** [comment]
All green.

**Comment by ericchiang** [comment]
woot!

**Pull request merged**